File: 3486.longest-special-path-ii.cpp
#
# @lc app=leetcode id=3486 lang=cpp
#
# [3486] Longest Special Path II
#

# @lc code=start
class Solution {
public:
    vector<int> longestSpecialPath(vector<vector<int>>& edges, vector<int>& nums) {
        int n = nums.size();
        vector<vector<pair<int, int>>> adj(n);
        for (auto& edge : edges) {
            adj[edge[0]].emplace_back(edge[1], edge[2]);
            adj[edge[1]].emplace_back(edge[0], edge[2]);
        }
        
        int maxLength = 0;
        int minNodes = INT_MAX;
        unordered_map<int, int> nodeValueCounts;
        vector<bool> visited(n, false);

        function<void(int, int, int)> dfs = [&](int node, int length, int nodesCount) {
            nodeValueCounts[nums[node]]++;
            visited[node] = true;
            
            int duplicateCount = 0;
            for (const auto& p : nodeValueCounts) {
                if (p.second > 1) duplicateCount++;
            }
            if (duplicateCount <= 1) {
                if (length > maxLength || (length == maxLength && nodesCount < minNodes)) {
                    maxLength = length;
                    minNodes = nodesCount;
                }
                for (auto& [neighbor, weight] : adj[node]) {
                    if (!visited[neighbor]) {
                        dfs(neighbor, length + weight, nodesCount + 1);
                    }
                }
            }
            
nodenodeValueCounts[nums[node]]--;
nodenodeif (nodeValueCounts[nums[node]] == 0) {
nodenodenodeValueCounts.erase(nums[node]);
nodenodene}
nodenodevisited[node] = false;
odendfs(0 /*root*/, 0 /*path length*/, 1 /*starting with root*/);
odreturn {maxLength,minNodes};od}}; # @lc code=end

File: 3486.longest-special-path-ii.go
#
# @lc app=leetcode id=3486 lang=golang
#
# [3486] Longest Special Path II
#

# @lc code=start
func longestSpecialPath(edges [][]int, nums []int) []int {
    type Node struct {
        id int
        length int
    }
    n := len(nums)
    graph := make([][]Node, n)
    for _, edge := range edges {
        u, v, length := edge[0], edge[1], edge[2]
        graph[u] = append(graph[u], Node{id: v, length: length})
        graph[v] = append(graph[v], Node{id: u, length: length})
    }
    
    var dfs func(node int, parent int, path map[int]int) (int, int)
    dfs = func(node int, parent int, path map[int]int) (int, int) {
        maxPathLength := 0
        minNodes := n // Start with max nodes possible for comparison
        
        newPath := make(map[int]int)
        for k, v := range path {
            newPath[k] = v
        }
        newPath[nums[node]]++
      
        distinctCount := len(newPath)
        duplicates := 0 // Track duplicates separately with a counter
        for _, count := range newPath {
            if count > 1 {
                duplicates++
            }
        }
      
        if duplicates > 1 { // More than one duplicate invalidates this path branch
            return 0, n // Invalid path returns zero length and max nodes (to be ignored)
        }
      
        for _, neighbor := range graph[node] {
            if neighbor.id == parent {
                continue // Avoid revisiting parent node in DFS
            }
            
            length, nodes := dfs(neighbor.id, node, newPath)
          totalLength := length + neighbor.length // Accumulate edge lengths in path computation
          if totalLength > maxPathLength || (totalLength == maxPathLength && nodes < minNodes) {
              maxPathLength = totalLength
              minNodes = nodes + 1 // Include current node in count
          }
      }
      return maxPathLength, minNodes
   }
   
   resultLength := 0
   resultNodes := n // Start with max nodes possible for initialization comparison purposes
   
   length, nodes := dfs(0, -1, map[int]int{}) // Start DFS from root node (node 0)
   if length > resultLength || (length == resultLength && nodes < resultNodes) {
       resultLength = length
       resultNodes = nodes
   }
   return []int{resultLength, resultNodes}
in}
in# @lc code=end